core/src/main/java/io/questdb/cairo/O3MemoryPressureRegulatorImpl.java

```diff
@@ -0,0 +1,160 @@
+ * when the system is under memory pressure.
+ * <p>
+ * This class is NOT thread-safe.
+ */
+public final class O3MemoryPressureRegulatorImpl implements O3MemoryPressureRegulator {
+    private static final Log LOG = LogFactory.getLog(O3MemoryPressureRegulatorImpl.class);
+
+    private static final int MAX_LEVEL = 10;
+    private static final int MIN_LEVEL = 0;
+    private static final int PARALLELISM_THROTTLING_LEVEL = 5; // when exceeded we start introducing back off
+    private final Rnd rnd;
+    private final SeqTxnTracker txnTracker;
+    private int level;
+    private long walBackOff = -1;
+
+    public O3MemoryPressureRegulatorImpl(Rnd rnd, MicrosecondClock clock, SeqTxnTracker txnTracker) {
+        this.rnd = rnd;
+        this.txnTracker = txnTracker;
+        this.level = txnTracker.getMemoryPressureLevel();
+        adjustWalBackOff(clock.getTicks());
+    }
+
+    @TestOnly
+    public int getLevel() {
+        return level;
+    }
+
+    @Override
+    public int getMaxO3MergeParallelism() {
+        if (level == 0) {
+            // fast path
+            return Integer.MAX_VALUE;
+        }
+
+        int adjustedLevel = Math.min(level, PARALLELISM_THROTTLING_LEVEL) * 10;
+        int partitionParallelism = 51 - adjustedLevel;
+        assert partitionParallelism > 0;
+        // level 1 = parallelism 41
+        // level 2 = parallelism 31
+        // level 3 = parallelism 21
+        // level 4 = parallelism 11
+        // level 5 = parallelism 1
+        // level 6 = parallelism 1 + backoff
+        // level 7 = parallelism 1 + backoff
+        // level 8 = parallelism 1 + backoff
+        // level 9 = parallelism 1 + backoff
+        // level 10 = parallelism 1 + backoff
+        return partitionParallelism;
+    }
+
+    @Override
+    public void onPressureDecreased(long nowMicros) {
+        if (level == MIN_LEVEL) {
+            // fast path
+            return;
+        }
+
+        if (level > PARALLELISM_THROTTLING_LEVEL) {
+            level--;
+            LOG.infoW().$("Memory pressure building up, new level=").$(level).$();
+            txnTracker.setMemoryPressureLevel(level);
+            adjustWalBackOff(nowMicros);
+        } else {
+            assert walBackOff == -1; // no backOff
+
+            // we increase parallelism probabilistically - the more pressure the lesser chance of increasing it
+            int minTrailingZeros = (level + 2);
+            // if memoryPressure == 1  we require 3 trailing zeros -> that's 1 in 8 chance
+            // if memoryPressure == 2  we require 4 trailing zeros -> that's 1 in 16 chance
+            // ...
+            // if memoryPressure == 5 we require 7 trailing zeros -> that's 1 in 128 chance
+            // yes, we are assuming that random number generator is reasonably good
+            int i = rnd.nextInt();
+            if (Integer.numberOfTrailingZeros(i) >= minTrailingZeros) {
```

Comment:
nit: I think this is has better readability of the probability you described:
```suggestion
            if (rnd.nextInt(Math.pow(2, Math.max(level, PARALLELISM_THROTTLING_LEVEL) +2)) == 1) {
```

Reply:
is it? 😮 
@mtopolnik wdyt?

Reply:
The proposed one-liner isn't more readable to me, but perhaps if we expand it and take the main idea to use an explicit range instead of number of trailing zeros, it may be a bit better. I wouldn't use math.pow though, but left shift.